httpcore/_backends/asyncio.py

```diff
@@ -78,6 +85,50 @@ async def backport_start_tls(
     return ssl_protocol._app_transport
 
 
+class DatagramProtocol(asyncio.DatagramProtocol):
+    def connection_lost(self, exc: Optional[Exception]) -> None:
+        self.close_event.set()
+
+    def connection_made(self, transport: asyncio.BaseTransport) -> None:
+        self.close_event = asyncio.Event()
+        self.read_queue: Deque[Tuple[bytes, SocketAddr]] = deque()
+        self.read_event = asyncio.Event()
+
+    def datagram_received(self, data: bytes, addr: SocketAddr) -> None:
+        self.read_queue.append((data, addr))
+        self.read_event.set()
+
+
+class DatagramSocket(AsyncDatagramSocket):
+    def __init__(
+        self,
+        transport: asyncio.DatagramTransport,
+        protocol: DatagramProtocol,
+        remote_addr: SocketAddr,
+    ):
+        self._protocol = protocol
+        self._transport = transport
+        self.remote_addr = remote_addr
+
+    async def aclose(self) -> None:
+        self._transport.close()
+        await self._protocol.close_event.wait()
+
+    async def receive(self) -> Tuple[bytes, SocketAddr]:
+        if not self._protocol.read_queue and not self._transport.is_closing():
+            await self._protocol.read_event.wait()
+
+        data, addr = self._protocol.read_queue.popleft()
+
+        if not self._protocol.read_queue:
+            self._protocol.read_event.clear()
+
+        return data, addr
+
+    async def send(self, data: bytes, addr: SocketAddr) -> None:
```

Comment:
Is there something we should be "waiting" for? In other backends this calls an async operation, IDK if asyncio exposes any kind of equivalent "actually wait for something to happen" API, instead of just scheduling something for later?

Reply:
To my knowledge UDP APIs don't provide any kind of backpressure.